File: tcp-proxy/src/s3-direct-put-object.js
const crypto = require('crypto');
// const fetch = require('node-fetch');
// const superagent = require('superagent');

// Replace with your AWS credentials and region
const AWS_ACCESS_KEY = process.env.AWS_ACCESS_KEY;
const AWS_SECRET_KEY = process.env.AWS_SECRET_KEY;
const AWS_REGION = 'eu-west-1'; // For example, 'us-west-2'

const bucketName = 'some-test';
const objectKey = 'example-object.txt';
const data = 'Hello, this is the content of the uploaded object!';

function getSignatureKey(key, date, region, service) {
  const kDate = crypto.HmacSHA256(date, 'AWS4' + key);
  const kRegion = crypto.HmacSHA256(region, kDate);
  const kService = crypto.HmacSHA256(service, kRegion);
  const kSigning = crypto.HmacSHA256('aws4_request', kService);
  return kSigning;
}

async function uploadObjectToBucket(bucketName, objectKey, data) {
  const method = 'PUT';
  const contentType = 'text/plain';
  const host = `${bucketName}.s3.amazonaws.com`;
  const currentDate = new Date().toISOString().replace(/[:\-]|\.\d{3}/g, '');
  const date = currentDate.slice(0, 8);

  const canonicalRequest = `${method}\n/${objectKey}\n\nhost:${host}\nx-amz-content-sha256:${crypto.SHA256(
    data
  ).toString()}\nx-amz-date:${currentDate}\n\nhost;x-amz-content-sha256;x-amz-date\n${crypto.SHA256(
    data
  )}`;

  const credentialScope = `${date}/${AWS_REGION}/s3/aws4_request`;
  const stringToSign = `AWS4-HMAC-SHA256\n${currentDate}\n${credentialScope}\n${crypto.SHA256(
    canonicalRequest
  ).toString()}`;

  const signingKey = getSignatureKey(
    AWS_SECRET_KEY,
    date,
    AWS_REGION,
    's3'
  );
  const signature = crypto.HmacSHA256(stringToSign, signingKey).toString();

  const url = `https://${host}/${objectKey}`;
  const headers = {
    'x-amz-content-sha256': crypto.SHA256(data).toString(),
    'x-amz-date': currentDate,
    Authorization: `AWS4-HMAC-SHA256 Credential=${AWS_ACCESS_KEY}/${credentialScope}, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=${signature}`,
    'Content-Type': contentType,
  };

  try {
    // await superagent
      
    const response = await fetch(url, {
      method,
      headers,
      body: data,
    });

    if (response.ok) {
      console.log(`Successfully uploaded object: ${objectKey}`);
    } else {
      console.error('Failed to upload object:', response.statusText);
    }
  } catch (error) {
    console.error('Error uploading object:', error.message);
  }
}

// Example usage:
uploadObjectToBucket(bucketName, objectKey, data);

File: tcp-proxy/src/s3-direct-put-object.py
import requests
import datetime
import hashlib
import hmac
import base64

# AWS credentials
aws_access_key = 'YOUR_ACCESS_KEY'
aws_secret_key = 'YOUR_SECRET_KEY'

# S3 bucket and object information
bucket_name = 'your-bucket-name'
object_key = 'your-object-key'

# S3 endpoint URL for your AWS region
s3_endpoint = 'https://s3.amazonaws.com'  # Update with the correct endpoint for your region

# HTTP method and content type
http_method = 'GET'
content_type = ''

# Current timestamp for the request
timestamp = datetime.datetime.utcnow().strftime('%Y%m%dT%H%M%SZ')

# Generate the canonical request
canonical_request = f"{http_method}\n/{bucket_name}/{object_key}\n\nhost:{s3_endpoint.replace('https://','')}\n\nhost\n"

# Generate the string to sign
string_to_sign = f"AWS4-HMAC-SHA256\n{timestamp}\n{timestamp[0:8]}/{aws_region}/s3/aws4_request\n{hashlib.sha256(canonical_request.encode('utf-8')).hexdigest()}"

# Generate the signing key
signing_key = hmac.new(("AWS4" + aws_secret_key).encode('utf-8'), timestamp[0:8].encode('utf-8'), hashlib.sha256)
signing_key = hmac.new(signing_key.digest(), aws_region.encode('utf-8'), hashlib.sha256)
signing_key = hmac.new(signing_key.digest(), b's3', hashlib.sha256)
signing_key = hmac.new(signing_key.digest(), b'aws4_request', hashlib.sha256)

# Calculate the signature
signature = hmac.new(signing_key.digest(), string_to_sign.encode('utf-8'), hashlib.sha256).hexdigest()

# Construct the headers
headers = {
    'Authorization': f"AWS4-HMAC-SHA256 Credential={aws_access_key}/{timestamp[0:8]}/{aws_region}/s3/aws4_request, SignedHeaders=host, Signature={signature}",
    'x-amz-date': timestamp,
}

# Make the request
response = requests.get(f"{s3_endpoint}/{bucket_name}/{object_key}", headers=headers)

# Handle the response
if response.status_code == 200:
    print("Object fetched successfully.")
    print(response.text)
else:
    print(f"Error: {response.status_code} - {response.text}")

